#!/usr/bin/env Rscript

###############################################################################

library(getopt);
library(parallel);

params=c(
	"vcf_list_fn", "v", 1, "character",
	"snp_loci_fn", "s", 1, "character",
	"output_dir", "o", 1, "character",
	"padding", "p", 2, "numeric"
);

opt=getopt(spec=matrix(params, ncol=4, byrow=TRUE), debug=FALSE);
script_name=unlist(strsplit(commandArgs(FALSE)[4],"=")[1])[2];

BCFTOOL_BIN="/usr/local/bin/bcftools";
DEF_PADDING=1000;

options(width=129);

usage = paste(
	"\nUsage:\n", script_name, "\n",
	"	-v <List of VCF Files (may be .gz)>\n",
	"	-s <File containing list of SNP loci>\n",
	"	-o <output directory>\n",
	"	[-p <Padding around target, default=", DEF_PADDING, "\n",
	"\n",
	"This tool depends on the bcftool.\n",
	"  The location it is looking for it is: ", BCFTOOL_BIN, "\n",
	"\n",
	"This script will read in a list of SNPs and then\n",
	"for each VCF file:\n",
	"	1.) Extract variants 1kb up and downstream of the location\n",
	"	2.) Extract the variant at the location\n",
	"	3.) The hom/het-ness of at the location\n",
	"	4.) The quality, if a variant\n",
	"	5.) The number of variants and mean depth around the location\n",
	"\n",
	"The input file should have been generated by:\n",
	"  QueryEnseml_dbSNP_by_rsID.r\n",
	"The first column should have the format of:\n",
	"     <chr#>:<start>-<end>\n",
	"\n",	
	"\n",
	"\n");

if(
	!length(opt$vcf_list_fn) || 
	!length(opt$snp_loci_fn) ||
	!length(opt$output_dir)
){
	cat(usage);
	q(status=-1);
}

InputVCFListFile=opt$vcf_list_fn;
SNPLociFile=opt$snp_loci_fn;
OutputDirectory=opt$output_dir;

Padding=DEF_PADDING;
if(length(opt$padding)){
	Padding=opt$padding;
}

cat("\n");
cat("Input VCF File List: ", InputVCFListFile, "\n");
cat("SNP Loci File: ", SNPLociFile, "\n");
cat("Output Dir: ", OutputDirectory, "\n");
cat("Padding: ", Padding, "\n");
cat("\n");

extraction_dir=paste(OutputDirectory, "/extract_regions", sep="");
target_genotype_dir=paste(OutputDirectory, "/target_genotype", sep="");
logs_dir=paste(OutputDirectory, "/logs", sep="");

dir.create(OutputDirectory);
dir.create(extraction_dir);
dir.create(target_genotype_dir);
dir.create(logs_dir);

##############################################################################
# Loading targeted SNPs
##############################################################################

cat("\n");
cat("Load targeted SNPs\n");
snp_table=read.table(SNPLociFile, header=T, sep="\t", quote="", 
	comment.char="#", stringsAsFactors=F);
num_snps=nrow(snp_table);

print(snp_table);
cat("Num SNPs Targeted: ", num_snps, "\n");
cat("\n");

snp_locs_cname=c("Chromosome", "Start", "End", "Target", "ExpRefAllele");
snp_locs=matrix(character(), nrow=num_snps, ncol=length(snp_locs_cname),
	dimnames=list(1:num_snps,snp_locs_cname));

for(i in 1:num_snps){
	toks=strsplit(snp_table[i,"chr_coord"], ":")[[1]];
	chrom=toks[1];
	toks=strsplit(toks[2], "-")[[1]];
	start=as.numeric(toks[1]);
	end=as.numeric(toks[2]);
	exp_ref_allele=strsplit(snp_table[i, "ref_var_alleles"], ",")[[1]][1];
	snp_locs[i, "Chromosome"]=paste("chr", chrom, sep="");
	snp_locs[i, "Start"]=max(1, start-Padding);
	snp_locs[i, "End"]=end+Padding;
	snp_locs[i, "Target"]=start;
	snp_locs[i, "ExpRefAllele"]=exp_ref_allele;
}
rownames(snp_locs)=chartr(":-","._",snp_table[,1]);

cat("SNPS reformatted and expanded for VCF Extraction:\n");
print(snp_locs);
cat("\n");

cat("Writing file for bcftool:\n");
snp_coords_vcf_targets_fn=paste(extraction_dir, "/vcf_targets.tsv", sep="");
write.table(x=snp_locs[,c("Chromosome", "Start", "End")], 
	file=snp_coords_vcf_targets_fn, sep="\t", quote=F,
	row.names=F, col.names=F);
cat(snp_coords_vcf_targets_fn, " written.\n", sep="");

##############################################################################
# Load VCF File
##############################################################################

cat("\n");
cat("Loading VCF File List:\n");
vcf_tab=read.table(InputVCFListFile, header=F, sep="\t", quote="",
	comment.char="#", stringsAsFactors=F);

vcf_list=vcf_tab[,1];
num_vcf_files=length(vcf_list);

cat("VCF Target List:\n");
print(vcf_list);
cat("Number of VCF Files Loaded: ", num_vcf_files, "\n");

##############################################################################
# Loop over VCF Files

extract_genotype=function(snp_loc_tab, genotype_tab_raw, padding){
	#cat("snp_loc_tab:\n");
	#print(snp_loc_tab);
	#cat("genotype_tab:\n");
	#print(genotype_tab);

	extract_col_from_info=function(info_col){
		# This will extract the allele freq and read depth from the info column
		# so it can be appended to position and allele table 

		flag_vals=strsplit(info_col, ";");

		num_rows=length(info_col);
		df=data.frame(nrow=num_rows, ncol=2);
		colnames(df)=c("AlleleFreq", "ReadDepth");

		for(i in 1:num_rows){
			#print(info_col[i]);
			flags=flag_vals[[i]];
			num_flags=length(flags);
			for(fix in 1:num_flags){

				curflag=flags[fix];
				if(length(grep("^AF=", curflag))){
					extracted=gsub("^AF=(\\d+\\.*\\d*)", "\\1", curflag);
					if(length(extracted)){
						df[i, "AlleleFreq"]=extracted;
					}
				}else if (length(grep("^DP=", curflag))){
					extracted=gsub("^DP=(\\d+)", "\\1", curflag);
					if(length(extracted)){
						df[i, "ReadDepth"]=extracted;
					}
				}

			}
			#print(df[i,]);
		}
		
		return(df);
	}

	genotype_tab=cbind(
		genotype_tab_raw[,setdiff(colnames(genotype_tab_raw), "info"),drop=F],
		extract_col_from_info(genotype_tab_raw[,"info"]));

	# Conver positions to numeric so we can search for it
	genotype_pos=as.numeric(genotype_tab[,"pos"]);

	num_snps_loc=nrow(snp_loc_tab);

	report_headers=c("Chromosome", "Pos", "Ref", "Var", "VarFreq", "Qual", 
		"MeanRegionQual", "NumRegionVariants");

	report_tab=matrix(character(), ncol=length(report_headers), nrow=num_snps_loc);
	colnames(report_tab)=report_headers;

	for(i in 1:num_snps_loc){
		cur_chrom=snp_loc_tab[i,"Chromosome"];
		cur_pos=as.numeric(snp_loc_tab[i,"Target"]);

		cat("Working on: \n");
		print(snp_loc_tab[i,]);

		chrom_subset_ix=genotype_tab[,"chrom"]==cur_chrom;
		genotype_tab_chrom=genotype_tab[chrom_subset_ix,,drop=F];
		genotype_pos_chrom=genotype_pos[chrom_subset_ix];
		chrom_range_ix=(genotype_pos_chrom<=(cur_pos+padding) & 
				genotype_pos_chrom>=(cur_pos-padding));
		inrange=genotype_tab_chrom[chrom_range_ix,,drop=F];

		cat("In Range Table: \n");
		print(inrange);

		num_region_variants=nrow(inrange);
		mean_region_qual=round(mean(inrange[,"qual"]), 2);
		# If there is an exact position, take it
		exact_ix=(cur_pos==inrange[,"pos"]);
		if(any(exact_ix)){
			cat("Exact positiong found!!!\n");
			inrange_ix=which(exact_ix);			
			record=inrange[inrange_ix,,drop=F];
			report_tab[i,
				c("Chromosome", "Pos", "Ref", "Var", "VarFreq", "Qual")]=c(
					record[1, "chrom"], 
					record[1, "pos"], 
					record[1, "ref"], 
					record[1, "alt"],
					record[1, "AlleleFreq"],
					record[1, "qual"]);
		}else{
			report_tab[i,
				c("Chromosome", "Pos", "Ref", "Var", "VarFreq", "Qual")]=c(
					cur_chrom,
					cur_pos,
					"",
					"",
					0,
					mean_region_qual
				);
		}

		report_tab[i, c("MeanRegionQual", "NumRegionVariants")]=c(
			mean_region_qual, num_region_variants);

		cat("\n\n\n");
	}

	#print(report_tab, quote=F);
	return(report_tab);
}

#vcf_tab=read.table("/home/kelvinli/git/AnalysisTools/VCF/Extract_Variants_by_Loci/example/results/extracted/30048_MOR107A81_CL_S18_001.vcf.extr_subset.info", header=F);
#colnames(vcf_tab)=c("chrom", "pos", "ref", "alt", "qual", "info");
#extract_genotype(snp_locs, vcf_tab, Padding);
#quit();

extract_region_vcf=function(vcf_file){
	vcf_file_path_tok=strsplit(vcf_file, "/")[[1]]
	vcf_fname=vcf_file_path_tok[length(vcf_file_path_tok)];
	vcf_file_root=gsub(".vcf.gz", "", vcf_fname);
	vcf_subset_info_file_root=paste(vcf_file_root, ".subset.info", sep="");

	vcf_subset_full_path=paste(extraction_dir,"/", vcf_subset_info_file_root, sep="");

	vcf_file_size=file.info(vcf_file)$size;
	if(vcf_file_size==0){
		return(paste("VCF File is zero lengthed: ", vcf_file, "\n", sep=""));
	}

	cmd=paste(
		BCFTOOL_BIN, " query ", vcf_file, 
		" -R ", snp_coords_vcf_targets_fn, 
		" -o ", vcf_subset_full_path, 
		" -f '%CHROM\t%POS\t%REF\t%ALT\t%QUAL\t%INFO'", sep="");
	
	log_txt=system(cmd, intern=T);
	log_full_path=paste(logs_dir,"/",vcf_file_root,".log.txt", sep="");
	write.table(log_txt, log_full_path, quote=F, row.names=F, col.names=F);

	subset_file_size=file.info(vcf_subset_full_path)$size;
	if(subset_file_size==0){
		return(paste("No targeted variants found: ", vcf_file, "\n", sep=""));
	}

	# Look up SNPs and figure out Variant, HomHet, etc.
	genotype_tab=read.table(file=vcf_subset_full_path, header=F);
	colnames(genotype_tab)=c("chrom", "pos", "ref", "alt", "qual", "info");
	indiv_report=extract_genotype(snp_locs, genotype_tab, Padding);

	genotype_full_path=paste(target_genotype_dir,"/",vcf_file_root,".genotype.tsv", sep="");
	write.table(indiv_report, genotype_full_path, quote=F, sep="\t", row.names=F,
		col.names=T);

	return(log_txt);
}

#num_cores=4;
num_cores=50;

cl=makeCluster(num_cores);
print(as.list(vcf_list));

clusterExport(cl, "BCFTOOL_BIN");
clusterExport(cl, "snp_coords_vcf_targets_fn");
clusterExport(cl, "extraction_dir");
clusterExport(cl, "target_genotype_dir");
clusterExport(cl, "logs_dir");
clusterExport(cl, "extract_genotype");
clusterExport(cl, "snp_locs");
clusterExport(cl, "Padding");

parLapply(cl, as.list(vcf_list), extract_region_vcf);

stopCluster(cl)

##############################################################################

cat("\nDone.\n");

##############################################################################

print(warnings());
q(status=0);
